File: ConcatenatedBinary.py
class Solution(object):
    def concatenatedBinary(self, n):
        """
        :type n: int
        :rtype: int
        """
        MOD = 10**9 + 7
        res = 0
        bit_length = 0
        
        for i in range(1, n + 1):
            # If i is a power of 2, the number of bits increases
            # (i & (i - 1)) == 0 is a quick way to check for powers of 2
            if i & (i - 1) == 0:
                bit_length += 1
            
            # Shift the result and add the current number
            res = ((res << bit_length) | i) % MOD
            
        return res

File: ConcatenatedBinary.java
class Solution {
    public int concatenatedBinary(int n) {
        long res = 0;
        int MOD = 1_000_000_007;
        int bitLength = 0;

        for (int i = 1; i <= n; i++) {
            // Check if i is a power of 2 to increment bitLength
            // e.g., 1 (1), 2 (10), 4 (100), 8 (1000)...
            if ((i & (i - 1)) == 0) {
                bitLength++;
            }

            // (res << bitLength) makes room for the new bits
            // | i adds the binary representation of i to the end
            res = ((res << bitLength) | i) % MOD;
        }

        return (int) res;
    }
}
